lib/graphql/schema/printer.rb

```diff
@@ -0,0 +1,131 @@
+module GraphQL
+  # Used to convert your {GraphQL::Schema} to a GraphQL schema string
+  #
+  # @example print your schema to standard output
+  #   Schema = GraphQL::Schema.new(query: QueryType)
+  #   puts GraphQL::Schema::Printer.print_schema(Schema)
+  #
+  module Schema::Printer
+    extend self
+
+    # Return a GraphQL schema string for the defined types in the schema
+    # @param schema [GraphQL::Schema]
+    def print_schema(schema)
+      print_filtered_schema(schema, method(:is_defined_type))
+    end
+
+    # Return the GraphQL schema string for the introspection type system
+    def print_introspection_schema
+      query_root = ObjectType.define do
+        name "Query"
+      end
+      schema = Schema.new(query: query_root)
+      print_filtered_schema(schema, method(:is_introspection_type))
+    end
+
+    private
+
+    def print_filtered_schema(schema, type_filter)
+      types = schema.types.values.select{ |type| type_filter.call(type) }.sort_by(&:name)
+      types.map{ |type| print_type(type) }.join("\n\n")
+    end
+
+    BUILTIN_SCALARS = Set.new(["String", "Boolean", "Int", "Float", "ID"])
+    private_constant :BUILTIN_SCALARS
+
+    def is_introspection_type(type)
+      type.name.start_with?("__")
+    end
+
+    def is_defined_type(type)
+      !is_introspection_type(type) && !BUILTIN_SCALARS.include?(type.name)
+    end
+
+    def print_type(type)
+      case type
```

Comment:
I'm not a big fan of giant `case` statements, I prefer a strategy class-based approach. Are there advantages to this style that I haven't considered? What do you think of applying that pattern here?

Here's one application of the strategy pattern:

- Setup a mapping of objects -> strategies: [BaseExecution#TYPE_KIND_STRATEGIES ](https://github.com/rmosolgo/graphql-ruby/blob/master/lib/graphql/query/base_execution/value_resolution.rb#L78-L85)
- When you receive an object, find the matching strategy and use it: [BaseResolution#result](https://github.com/rmosolgo/graphql-ruby/blob/master/lib/graphql/query/serial_execution/field_resolution.rb#L36-L38)

In short, you spread API surface area over a bunch of classes instead of adding them to one class. I like it because:

- Each class's API is smaller. In a class with a big API, it's harder to tell which methods get called where. In a class with a small API, it's easier to find all the possibilities.
- Any state is private. Each class can have its own instance variables without fear of messing with the others. (Doesn't apply here, but, in general!)
- A strategy-as-a-class is very expressive. Since the strategy is just a class, it can do anything a class can do (include, extend, define methods, etc), in the privacy of its own scope.

What do you think of applying that pattern here?